### PR TITLE
feat(optimizer): Add get_job_logs API to OptimizerClient

### PR DESCRIPTION
**Pull request by kramaranya**
I've added `get_job_logs` API to the OptimizerClient

Working example:
```python
from kubeflow.optimizer import OptimizerClient, Search, Objective, TrialConfig

def get_torch_dist(learning_rate: str, num_epochs: str):
    import os
    import torch
    import torch.distributed as dist

    dist.init_process_group(backend="gloo")
    print("PyTorch Distributed Environment")
    print(f"WORLD_SIZE: {dist.get_world_size()}")
    print(f"RANK: {dist.get_rank()}")
    print(f"LOCAL_RANK: {os.environ['LOCAL_RANK']}")

    lr = float(learning_rate)
    epochs = int(num_epochs)
    loss = 1.0 - (lr * 2) - (epochs * 0.01)

    if dist.get_rank() == 0:
        print(f"loss={loss}")
    
    dist.barrier()

template = TrainJobTemplate(
    trainer=CustomTrainer(
        func=get_torch_dist,
        func_args={"learning_rate": "0.01", "num_epochs": "5"},
        num_nodes=2,
        resources_per_node={"gpu": 1},
    ),
    runtime=TrainerClient().get_runtime("torch-distributed"),
)

job_id = OptimizerClient().optimize(
    trial_template=template,
    trial_config=TrialConfig(num_trials=10, parallel_trials=2),
    search_space={
        "learning_rate": Search.loguniform(0.001, 0.1),
        "num_epochs": Search.choice([5, 10, 15]),
    },
)

print(f"OptimizationJob created: {job_id}")

print("\n".join(OptimizerClient().get_job_logs(name=job_id)))
```
/assign @kubeflow/kubeflow-sdk-team

### CODE REVIEW

File: kubeflow/optimizer/api/optimizer_client.py [hunk 1]
```diff
@@ -119,6 +121,47 @@ def get_job(self, name: str) -> OptimizationJob:
 
         return self.backend.get_job(name=name)
 
+    def get_job_logs(
+        self,
+        name: str,
+        trial: Optional[str] = None,
```
Comment (andreyvelich) [inline kubeflow/optimizer/api/optimizer_client.py]: Can we use `trial_name` here ?
```suggestion
        trial_name: Optional[str] = None,
```

File: kubeflow/optimizer/backends/base.py [hunk 2]
```diff
@@ -47,6 +48,16 @@ def list_jobs(self) -> list[OptimizationJob]:
     def get_job(self, name: str) -> OptimizationJob:
         raise NotImplementedError()
 
+    @abc.abstractmethod
+    def get_job_logs(
+        self,
+        name: str,
+        trial: Optional[str],
```
Comment (andreyvelich) [inline kubeflow/optimizer/backends/base.py]:
```suggestion
        trial_name: Optional[str],
```

File: kubeflow/optimizer/backends/kubernetes/backend.py [hunk 3]
```diff
@@ -212,6 +213,60 @@ def get_job(self, name: str) -> OptimizationJob:
...
+        trial: Optional[str] = None,
+        follow: bool = False,
+        step: str = trainer_constants.NODE + "-0",
+    ) -> Iterator[str]:
+        """Get the OptimizationJob logs from a Trial"""
+        # Determine what trial to get logs from.
+        if trial is None:
+            # Get logs from the best current trial.
+            best_trial = self.get_best_trial(name)
+            if best_trial is None:
```
Comment (andreyvelich) [inline kubeflow/optimizer/backends/kubernetes/backend.py:227]: As we discussed, if the best Trial is empty, let's take the first Trial from the OptimizationJob if list is not empty: https://github.com/kubeflow/sdk/blob/63effe0ca460733e83524434dc947820aa646000/kubeflow/optimizer/types/optimization_types.py#L117

File: kubeflow/optimizer/backends/kubernetes/backend.py [hunk 4]
```diff
@@ -212,6 +213,60 @@ def get_job(self, name: str) -> OptimizationJob:
...
+                # TODO (kramaranya): Consider waiting for best trial when follow=True
+                return
+            trial = best_trial.name
+            logger.info(f"Getting logs from best trial: {trial}")
```
Comment (andreyvelich) [inline kubeflow/optimizer/backends/kubernetes/backend.py]: Let's use `.debug` here, since we don't use `.info` logging in the SDK for now.
```suggestion
            logger.debug(f"Getting logs from best trial: {trial}")
```

File: kubeflow/optimizer/backends/kubernetes/backend.py [hunk 5]
```diff
@@ -212,6 +213,60 @@ def get_job(self, name: str) -> OptimizationJob:
         optimization_job = self.__get_experiment_cr(name)
         return self.__get_optimization_job_from_cr(optimization_job)
 
+    def get_job_logs(
+        self,
+        name: str,
+        trial: Optional[str] = None,
+        follow: bool = False,
+        step: str = trainer_constants.NODE + "-0",
```
Comment (andreyvelich) [inline kubeflow/optimizer/backends/kubernetes/backend.py]: Maybe for now, we should remove step from this API ?
The problem is that for other Steps (e.g. Pods), the container name is not `metrics-logger-and-collector`.
As a workaround, users can always use `TrainerClient()` to get logs for other steps of TrainJob
(e.g. TrainJob name == Trial name)

File: kubeflow/optimizer/backends/kubernetes/backend.py [hunk 6]
```diff
@@ -212,6 +213,60 @@ def get_job(self, name: str) -> OptimizationJob:
...
+        pod_name = None
+        for c in self.trainer_backend.get_job(trial).steps:
+            if c.status != trainer_constants.POD_PENDING and c.name == step:
+                pod_name = c.pod_name
+                break
+        if pod_name is None:
+            return
+
+        container_name = constants.METRICS_COLLECTOR_CONTAINER
+        try:
```
Comment (andreyvelich) [inline kubeflow/optimizer/backends/kubernetes/backend.py]: To reduce code duplication can you wrap this code under helper function: `self.__read_pod_logs(pod_name: str, container_name: str, follow: bool) in `, the Trainer client and use it as:
```py
yield from self.trainer_backend.__read_pod_logs(
        pod_name=pod_name,
        container_name=container_name,
        follow=follow
    )
```

File: kubeflow/optimizer/backends/kubernetes/backend.py (hunk 3 above)
Comment (kramaranya) [reply kubeflow/optimizer/backends/kubernetes/backend.py:227]: Sure, updated in https://github.com/kubeflow/sdk/pull/148/commits/66927dc6affc44cd9c74ae16c91d642c3ea02ae8

File: kubeflow/optimizer/backends/kubernetes/backend.py (hunk 5 above)
Comment (kramaranya) [reply kubeflow/optimizer/backends/kubernetes/backend.py]: I was thinking about this too, makes sense to me, updated in https://github.com/kubeflow/sdk/pull/148/commits/e1a00b821e2b87ce688e9d3d9aa1e85982184c67

File: kubeflow/optimizer/backends/kubernetes/backend.py (hunk 6 above)
Comment (kramaranya) [reply kubeflow/optimizer/backends/kubernetes/backend.py]: Updated in https://github.com/kubeflow/sdk/pull/148/commits/49b4da0e00280e988f79d3bb9fa1968c6a9434f9

**Comment by astefanutti** [comment]
Thanks @kramaranya 

/lgtm
/approve

I think we need to improve how Katib instruments the train nodes for StdOutCollector and FileCollector metrics to avoid depending on the metric collector sidecar and also make it work for other configurations like TfEventCollector or PrometheusMetricCollector.
